src/freenet/support/MultiValueTable.java

```diff
@@ -1,230 +1,260 @@
+    }
+
+    @SafeVarargs
+    public static <K, V> MultiValueTable<K, V> from(K key, V... values) {
+        return from(key, Arrays.asList(values));
+    }
+
+    public static <K, V> MultiValueTable<K, V> from(K key, Collection<V> values) {
+        MultiValueTable<K, V> table = new MultiValueTable<>(1);
+        table.putAll(key, values);
+        return table;
     }
 
     public void put(K key, V value) {
-        synchronized (table) {
-            Vector<V> v = table.get(key);
-            if (v == null) {
-                v = new Vector<V>(ies);
-                table.put(key, v);
+        this.table.compute(key, (k, list) -> {
+            if (list == null) {
+                list = new CopyOnWriteArrayList<>();
             }
-            v.addElement(value);
-        }
+            list.add(value);
+            return list;
+        });
+    }
+
+    public void putAll(K key, Collection<V> elements) {
+        this.table.compute(key, (k, list) -> {
+            if (list == null) {
+                list = new CopyOnWriteArrayList<>(elements);
+            } else {
+                list.addAll(elements);
+            }
+            return list;
+        });
     }
 
     /**
      * Returns the first element for this key.
      */
     public V get(K key) {
-        synchronized (table) {
-            Vector<V> v = table.get(key);
-            return (v == null ?
-                    null :
-                    v.firstElement());
+        return getFirst(key);
+    }
+
+    /**
+     * Returns the first element for this key.
+     */
+    public V getFirst(K key) {
+        List<V> list = this.table.get(key);
+        if (list == null || list.isEmpty()) {
+            return null;
         }
+        return list.get(0);
     }
 
     public boolean containsKey(K key) {
-		synchronized (table) {
-			return table.containsKey(key);
-		}
+        return this.table.containsKey(key);
     }
 
     public boolean containsElement(K key, V value) {
-        synchronized (table) {
-            Vector<V> v = table.get(key);
-            return (v != null) && v.contains(value);
+        List<V> list = this.table.get(key);
+        if (list == null || list.isEmpty()) {
+            return false;
         }
+        return list.contains(value);
     }
 
     /**
-     * Users will have to handle synchronizing.
+     * Returns mapped value collection as {@link Enumeration}
+     *
+     * <b>Note:</b> this method is deprecated
+     * please use other {@link #getAllAsList(Object)} method variant, which provides a {@link List}
+     *
+     * @param key key mapping
      */
+    @Deprecated
     public Enumeration<V> getAll(K key) {
-    	Vector<V> v;
-		synchronized (table) {
-			v = table.get(key);
-		}
-        return (v == null ?
-                new EmptyEnumeration<V>() :
-                v.elements());
+        List<V> elements = getAllAsList(key);
+        if (elements.isEmpty()) {
+            return Collections.emptyEnumeration();
```

Comment:
This special case is unnecessary.

Reply:
The only benefit of `Collections.emptyEnumeration()` is that it returns same object instance all the time.
Otherwise, yes, it can be simpler.
Will make it to be one-liner then.